### PR TITLE
Api to fetch catalog errors

### PR DESCRIPTION
**Pull request by pratap0007**
This api returns all type of errors(warn,info,critical) of a catalog
which occurred while refreshing it and users need to have a
`catalog:refresh` scope to access this api

Endpoint: Get `/catalog/<catalog_name>/error`

Closes: #208 

Signed-off-by: Shiv Verma <shverma@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/main/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/pkg/service/category/category.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func (s *service) List(ctx context.Context) (*category.ListResult, error) {
 
 	log := s.Logger(ctx)
 	db := s.DB(ctx)
-
+
```
Comment (piyush-garg) [inline api/pkg/service/category/category.go:45]: can you please undo this

File: api/design/catalog.go [hunk 2]
```diff
@@ -69,4 +69,30 @@ var _ = Service("catalog", func() {
 		})
 	})
 
+	Method("CatalogError", func() {
+		Description("List the all type of errors of a catalog")
```
Comment (sm43) [inline api/design/catalog.go:73]: `List all errors occurred refreshing a catalog`?

File: api/design/catalog.go [hunk 3]
```diff
@@ -69,4 +69,30 @@ var _ = Service("catalog", func() {
...
+		Security(types.JWTAuth, func() {
+			Scope("catalog:refresh")
+		})
+		Payload(func() {
+			Attribute("catalogName", String, "Name of catalog", func() {
+				Example("catalogName", "tekton")
+			})
+			Token("token", String, "JWT")
```
Comment (sm43) [inline api/design/catalog.go:81]: `Example`? just to make sure openapi file doesn't have any random string as example

File: api/design/types/type.go [hunk 4]
```diff
@@ -410,3 +410,10 @@ var UserData = Type("UserData", func() {
 	})
 	Required("githubId", "name", "avatarUrl")
 })
+
+var CatalogErrors = Type("CatalogErrors", func() {
+	Description("CatalogErrors define the errors that occurred during catalog refresh")
```
Comment (sm43) [inline api/design/types/type.go:415]: `Examples`

File: api/gen/http/openapi3.yaml [hunk 5]
```diff
@@ -109,6 +103,70 @@ paths:
...
+          description: OK response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/CatalogErrorResponseBody'
+              example:
+                data:
+                - errors:
+                  - Itaque non aut qui.
+                  - Consequatur assumenda suscipit aut minima autem ut.
```
Comment (sm43) [inline api/gen/http/openapi3.yaml:134]: random strings will be added if no examples are there

File: api/pkg/service/catalog/catalog.go [hunk 6]
```diff
@@ -106,6 +107,50 @@ func (s *service) RefreshAll(ctx context.Context, p *catalog.RefreshAllPayload)
...
 }
 
+// List the all type of errors of a catalog
+func (s *service) CatalogError(ctx context.Context, p *catalog.CatalogErrorPayload) (*catalog.CatalogErrorResult, error) {
+
+	log := s.Logger(ctx)
+	db := s.DB(ctx)
+
+	ctg := model.Catalog{}
+	if err := db.Where(&model.Catalog{Name: p.CatalogName}).Model(&model.Catalog{}).First(&ctg).Error; err != nil {
```
Comment (sm43) [inline api/pkg/service/catalog/catalog.go:117]: `Model(&model.Catalog{})` is not required I guess. can you pls check without it does the query executes the same?

File: api/pkg/service/catalog/catalog.go [hunk 7]
```diff
@@ -106,6 +107,50 @@ func (s *service) RefreshAll(ctx context.Context, p *catalog.RefreshAllPayload)
...
+		if err == gorm.ErrRecordNotFound {
+			return nil, catalogNotFoundErr(p.CatalogName)
+		}
+		log.Error(err)
+		return nil, internalError
+	}
+
+	var allErrors []string
+
+	db.Model(&model.CatalogError{CatalogID: ctg.ID}).Distinct("type").Find(&allErrors)
```
Comment (sm43) [inline api/pkg/service/catalog/catalog.go:127]: what if we fetch by only one query and then segregate?
`select * from CatalogError  where catalog_id=<>`
Create 3 arrays for each type ..loop on catalogError array and add error in respective array
and if array is not empty -> add in response object?
wdyt?

File: api/pkg/service/catalog/catalog_test.go [hunk 8]
```diff
@@ -120,3 +120,23 @@ func TestRefresh_All(t *testing.T) {
...
+
+	catalogSvc := NewServiceTest(tc)
+	ctx := auth.WithUserID(context.Background(), user.ID)
+
+	payload := &catalog.CatalogErrorPayload{CatalogName: "catalog-official"}
+	res, err := catalogSvc.CatalogError(ctx, payload)
+	assert.NoError(t, err)
+
+	assert.Equal(t, "info", res.Data[0].Type)
+	assert.Equal(t, []string{"display name is missing for buildah task"}, res.Data[0].Errors)
```
Comment (sm43) [inline api/pkg/service/catalog/catalog_test.go:141]: could you create a golden file similar to we do for test in resource service?

File: api/pkg/service/catalog/catalog.go [hunk 9]
```diff
@@ -106,6 +112,53 @@ func (s *service) RefreshAll(ctx context.Context, p *catalog.RefreshAllPayload)
...
+		case warning:
+			rs[warning] = append(rs[warning], item.Detail)
+		case critical:
+			rs[critical] = append(rs[critical], item.Detail)
+		default:
+			rs[unknown] = append(rs[unknown], item.Detail)
+		}
+	}
+
+	ret := &catalog.CatalogErrorResult{}
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/catalog.go:151]: Can we have a better variable name ??

File: api/pkg/service/catalog/catalog.go [hunk 10]
```diff
@@ -106,6 +112,53 @@ func (s *service) RefreshAll(ctx context.Context, p *catalog.RefreshAllPayload)
...
+		case info:
+			rs[info] = append(rs[info], item.Detail)
+		case warning:
+			rs[warning] = append(rs[warning], item.Detail)
+		case critical:
+			rs[critical] = append(rs[critical], item.Detail)
+		default:
+			rs[unknown] = append(rs[unknown], item.Detail)
+		}
+	}
```
Comment (sm43) [inline api/pkg/service/catalog/catalog.go:149]:
```
for _, item := range allCtgError {
		switch item.Type {
		case info,warning,critical:
			rs[item.Type] = append(rs[item.Type], item.Detail)
		default:
			rs[unknown] = append(rs[unknown], item.Detail)
		}
}
```

File: api/pkg/service/catalog/testdata/TestCatalogError.golden [hunk 11]
```diff
@@ -0,0 +1 @@
+display name is missing for buildah task
```
Comment (sm43) [inline api/pkg/service/catalog/testdata/TestCatalogError.golden:1]: you can print the whole api response returned

Comment (sm43) [reply api/pkg/service/catalog/testdata/TestCatalogError.golden:1]: write a http test similar to https://github.com/tektoncd/hub/blob/6c97d9e8d2fda05a60d294e2759978c87dc57594/api/pkg/service/resource/resource_http_test.go#L239

File: api/pkg/service/catalog/catalog_http_test.go [hunk 12]
```diff
@@ -93,3 +104,27 @@ func TestRefreshAll_Http(t *testing.T) {
 		assert.Equal(t, 2, len(res))
 	})
 }
+
+func TestCatalogError_Http(t *testing.T) {
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/catalog_http_test.go:108]: Could you please add a test which has no errors with a golden file ????

File: api/pkg/service/catalog/catalog_test.go [hunk 13]
```diff
@@ -120,3 +120,24 @@ func TestRefresh_All(t *testing.T) {
 	assert.Contains(t, []string{"catalog-official", "catalog-community"}, jobs[1].CatalogName)
 	assert.Equal(t, "queued", jobs[1].Status)
 }
+
+func TestCatalogError(t *testing.T) {
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/catalog_test.go:124]: Could you please add a test here when the errors occurs and assert the error message ?

File: api/pkg/service/catalog/catalog_test.go [hunk 14]
```diff
@@ -120,3 +120,24 @@ func TestRefresh_All(t *testing.T) {
...
+	tc := testutils.Setup(t)
+	testutils.LoadFixtures(t, tc.FixturePath())
+
+	// user with catalog:refresh scope
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/catalog_test.go:128]: nit: 

```suggestion
	// User with catalog:refresh scope
```

File: api/pkg/service/catalog/catalog_http_test.go [hunk 15]
```diff
@@ -93,3 +104,27 @@ func TestRefreshAll_Http(t *testing.T) {
...
+	tc := testutils.Setup(t)
+	testutils.LoadFixtures(t, tc.FixturePath())
+
+	// user with catalog:refresh scope
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/catalog_http_test.go:112]: nit:

```suggestion
	// User with catalog:refresh scope
```

**Review by PuneetPunamiya — COMMENTED** [review]
Thanks for the pr!
Looks good, only a small couple of comments to address

File: api/pkg/service/catalog/testdata/TestCatalogError_HttpHavingNoError.golden [hunk 16]
```diff
@@ -0,0 +1,3 @@
+{
+	"data": null
```
Comment (PuneetPunamiya) [inline api/pkg/service/catalog/testdata/TestCatalogError_HttpHavingNoError.golden:2]: Could you please return an empty array instead of null ??

**Pull request merged**